File: tessandiocyle_keyx_c/reconcile.c
/**
 * TessanDioKey V18 — Tri-layer Reconciliation
 * Translated from Rust V18 reconcile.rs
 */
#include "tessandiocyle_keyx.h"
#include <string.h>

static uint8_t coarse_bucket(int16_t v)
{
    int32_t u = v;
    u += (u >> 31) & TD18_Q;
    int32_t scaled = (u * 8 * 2 + TD18_Q) / (2 * TD18_Q);
    int32_t bucket = (scaled * 8) / 9;
    if (bucket > 7) bucket = 7;
    return (uint8_t)bucket;
}

static uint8_t fine_sub(int16_t v, uint8_t coarse)
{
    int32_t u = v;
    u += (u >> 31) & TD18_Q;
    int32_t low  = (int32_t)coarse * TD18_Q / 8;
    int32_t high = ((int32_t)coarse + 1) * TD18_Q / 8;
    int32_t mid = (low + high) / 2;
    return (u >= mid) ? 1 : 0;
}

static uint8_t spectral_pair(const int16_t block[4])
{
    int32_t s0 = (int32_t)block[0] + block[1] - block[2] - block[3];
    int32_t s1 = (int32_t)block[0] + block[2] - block[1] - block[3];
    uint8_t b0 = (s0 >= 0) ? 1 : 0;
    uint8_t b1 = (s1 >= 0) ? 1 : 0;
    return (uint8_t)((b1 << 1) | b0);
}

void td18_reconcile_alice(uint8_t hints[TD18_RECON_BYTES], const int16_t v[TD18_N])
{
    memset(hints, 0, TD18_RECON_BYTES);
    const int cfb = 128;
    for (int i = 0; i < TD18_N; i++) {
        uint8_t c = coarse_bucket(v[i]);
        uint8_t f = fine_sub(v[i], c);
        uint8_t bits = (uint8_t)(((c & 7) << 1) | (f & 1));
        int bit_pos = i * 4;
        int byte_idx = bit_pos / 8;
        int bit_off  = bit_pos % 8;
        hints[byte_idx] |= bits << bit_off;
        if (bit_off + 4 > 8) hints[byte_idx + 1] |= bits >> (8 - bit_off);
    }
    for (int block_idx = 0; block_idx < 64; block_idx++) {
        int16_t block[4];
        for (int j = 0; j < 4; j++) block[j] = v[block_idx * 4 + j];
        uint8_t sp = spectral_pair(block);
        int bit_pos = block_idx * 2;
        int byte_idx = cfb + (bit_pos / 8);
        int bit_off  = bit_pos % 8;
        hints[byte_idx] |= sp << bit_off;
        if (bit_off + 2 > 8) hints[byte_idx + 1] |= sp >> (8 - bit_off);
    }
}

int td18_reconcile_verify(const int16_t va[TD18_N], const int16_t vb[TD18_N])
{
    const int32_t SPECTRAL_TOLERANCE = 64;
    for (int i = 0; i < TD18_N; i++) {
        uint8_t ca = coarse_bucket(va[i]);
        uint8_t cb = coarse_bucket(vb[i]);
        int diff = (int)ca - (int)cb;
        if (diff < 0) diff = -diff;
        if (diff > 1) return TD18_ERR_VERIFY;
    }
    for (int block_idx = 0; block_idx < 64; block_idx++) {
        int16_t a[4], b[4];
        for (int j = 0; j < 4; j++) { a[j] = va[block_idx * 4 + j]; b[j] = vb[block_idx * 4 + j]; }
        int32_t s0_a = (int32_t)a[0] + a[1] - a[2] - a[3];
        int32_t s0_b = (int32_t)b[0] + b[1] - b[2] - b[3];
        int32_t s1_a = (int32_t)a[0] + a[2] - a[1] - a[3];
        int32_t s1_b = (int32_t)b[0] + b[2] - b[1] - b[3];
        int32_t as0 = s0_a >= 0 ? s0_a : -s0_a;
        int32_t as0b = s0_b >= 0 ? s0_b : -s0_b;
        int32_t as1 = s1_a >= 0 ? s1_a : -s1_a;
        int32_t as1b = s1_b >= 0 ? s1_b : -s1_b;
        int s0_ok = ((s0_a >= 0 && s0_b >= 0) || (s0_a < 0 && s0_b < 0) || (as0 < SPECTRAL_TOLERANCE && as0b < SPECTRAL_TOLERANCE));
        int s1_ok = ((s1_a >= 0 && s1_b >= 0) || (s1_a < 0 && s1_b < 0) || (as1 < SPECTRAL_TOLERANCE && as1b < SPECTRAL_TOLERANCE));
        if (!s0_ok || !s1_ok) return TD18_ERR_VERIFY;
    }
    return TD18_OK;
}

File: tessandiocyle_keyx/src/reconcile.rs
// TessanDioKey v16 — Tri-layer reconciliation: coarse 3-bit + fine 1-bit + spectral 2-bit WHT
// Spectral layer catches correlated noise across 4-coeff blocks via Walsh-Hadamard parity.
use crate::params::{
    KYBER_N, KYBER_Q, RECONCILE_COARSE_BITS, RECONCILE_FINE_BITS, RECONCILE_Q_EFF,
    RECONCILE_SPECTRAL_BITS,
};

const COARSE_BUCKETS: i16 = 1 << RECONCILE_COARSE_BITS; // 8
#[allow(dead_code)]
const FINE_SUBDIV: i16 = 1 << RECONCILE_FINE_BITS;      // 2
const COFINE_BITS: usize = RECONCILE_COARSE_BITS + RECONCILE_FINE_BITS; // 4
const BLOCK_SIZE: usize = 4; // coefficients per spectral block
const BLOCK_COUNT: usize = KYBER_N / BLOCK_SIZE; // 64

/// Coarse bucket index for a coefficient (0..7).
#[inline]
fn coarse_bucket(v: i16) -> u8 {
    let mut u = v as i32;
    u += (u >> 15) & (KYBER_Q as i32);
    let scaled = (u * (RECONCILE_Q_EFF as i32) * 2 + (KYBER_Q as i32)) / (2 * (KYBER_Q as i32));
    let bucket = (scaled * (COARSE_BUCKETS as i32)) / (RECONCILE_Q_EFF as i32 + 1);
    bucket.min((COARSE_BUCKETS - 1) as i32) as u8
}

/// Fine sub-bucket bit for a coefficient (0 or 1).
#[inline]
fn fine_sub(v: i16, coarse: u8) -> u8 {
    let mut u = v as i32;
    u += (u >> 15) & (KYBER_Q as i32);
    let bucket_low = (coarse as i32) * (KYBER_Q as i32) / (COARSE_BUCKETS as i32);
    let bucket_high = ((coarse as i32) + 1) * (KYBER_Q as i32) / (COARSE_BUCKETS as i32);
    let mid = (bucket_low + bucket_high) / 2;
    if u >= mid { 1 } else { 0 }
}

/// Spectral layer: 2-bit Walsh-Hadamard parity over a 4-coeff block.
/// Block layout: [c0, c1, c2, c3].
///   bit0: sign of (c0 + c1) - (c2 + c3)  → captures vertical correlation
///   bit1: sign of (c0 + c2) - (c1 + c3)  → captures diagonal correlation
#[inline]
fn spectral_pair(block: &[i16; BLOCK_SIZE]) -> u8 {
    let s0 = block[0] as i32 + block[1] as i32 - block[2] as i32 - block[3] as i32;
    let s1 = block[0] as i32 + block[2] as i32 - block[1] as i32 - block[3] as i32;
    let b0 = if s0 >= 0 { 1u8 } else { 0u8 };
    let b1 = if s1 >= 0 { 1u8 } else { 0u8 };
    (b1 << 1) | b0
}

/// Number of bytes for coarse+fine layer.
pub fn cofine_bytes() -> usize {
    (KYBER_N * COFINE_BITS).div_ceil(8) // 128
}

/// Number of bytes for spectral layer.
pub fn spectral_bytes() -> usize {
    (BLOCK_COUNT * RECONCILE_SPECTRAL_BITS).div_ceil(8) // 16
}

/// Total reconciliation bytes (coarse+fine+spectral).
pub fn reconcile_total_bytes() -> usize {
    cofine_bytes() + spectral_bytes() // 144
}

/// Reconcile a polynomial into tri-layer packed bytes.
/// Layout: [cofine_layer (128 bytes)] || [spectral_layer (16 bytes)]
pub fn reconcile_alice(v: &[i16; KYBER_N]) -> Vec<u8> {
    let cfb = cofine_bytes();
    let spb = spectral_bytes();
    let mut out = vec![0u8; cfb + spb];

    // Coarse+fine layer
    for (i, &vi) in v.iter().enumerate() {
        let c = coarse_bucket(vi);
        let f = fine_sub(vi, c);
        let bits = ((c & 0x7) << 1) | (f & 0x1);
        let bit_pos = i * COFINE_BITS;
        let byte_idx = bit_pos / 8;
        let bit_off = bit_pos % 8;
        out[byte_idx] |= bits << bit_off;
        if bit_off + COFINE_BITS > 8 {
            out[byte_idx + 1] |= bits >> (8 - bit_off);
        }
    }

    // Spectral layer
    for block_idx in 0..BLOCK_COUNT {
        let mut block = [0i16; BLOCK_SIZE];
        for j in 0..BLOCK_SIZE {
            block[j] = v[block_idx * BLOCK_SIZE + j];
        }
        let sp = spectral_pair(&block);
        let bit_pos = block_idx * RECONCILE_SPECTRAL_BITS;
        let byte_idx = cfb + (bit_pos / 8);
        let bit_off = bit_pos % 8;
        out[byte_idx] |= sp << bit_off;
        if bit_off + RECONCILE_SPECTRAL_BITS > 8 {
            out[byte_idx + 1] |= sp >> (8 - bit_off);
        }
    }

    out
}

/// Bob computes the same tri-layer reconciliation data from his view.
pub fn reconcile_bob_hints(v: &[i16; KYBER_N]) -> Vec<u8> {
    reconcile_alice(v)
}

/// Extract reconciled shared bits.
/// In this design the hints themselves are the reconciled bits (both parties
/// agree with overwhelming probability because noise << bucket width).
pub fn reconcile_extract(_v_alice: &[i16; KYBER_N], hints: &[u8]) -> Vec<u8> {
    hints.to_vec()
}

/// Verify tri-layer reconciliation succeeds with tolerance.
/// Coarse buckets must match within ±1.
/// Spectral parity must match unless both underlying sums are near zero
/// (within threshold), in which case noise may legitimately cross the boundary.
pub fn reconcile_verify(v_alice: &[i16; KYBER_N], v_bob: &[i16; KYBER_N]) -> bool {
    const SPECTRAL_TOLERANCE: i32 = 64; // 4 * max_expected_noise ≈ 4*16

    // Coarse+fine verification
    for i in 0..KYBER_N {
        let ca = coarse_bucket(v_alice[i]);
        let cb = coarse_bucket(v_bob[i]);
        if ca.abs_diff(cb) > 1 {
            return false;
        }
    }
    // Spectral parity verification with near-zero tolerance
    for block_idx in 0..BLOCK_COUNT {
        let mut a = [0i16; BLOCK_SIZE];
        let mut b = [0i16; BLOCK_SIZE];
        for j in 0..BLOCK_SIZE {
            a[j] = v_alice[block_idx * BLOCK_SIZE + j];
            b[j] = v_bob[block_idx * BLOCK_SIZE + j];
        }
        let s0_a = a[0] as i32 + a[1] as i32 - a[2] as i32 - a[3] as i32;
        let s0_b = b[0] as i32 + b[1] as i32 - b[2] as i32 - b[3] as i32;
        let s1_a = a[0] as i32 + a[2] as i32 - a[1] as i32 - a[3] as i32;
        let s1_b = b[0] as i32 + b[2] as i32 - b[1] as i32 - b[3] as i32;

        let s0_ok = s0_a.signum() == s0_b.signum()
            || (s0_a.abs() < SPECTRAL_TOLERANCE && s0_b.abs() < SPECTRAL_TOLERANCE);
        let s1_ok = s1_a.signum() == s1_b.signum()
            || (s1_a.abs() < SPECTRAL_TOLERANCE && s1_b.abs() < SPECTRAL_TOLERANCE);
        if !s0_ok || !s1_ok {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::Rng;

    #[test]
    fn test_tri_layer_reconcile_identity() {
        let v = [1500i16; KYBER_N];
        let a = reconcile_alice(&v);
        let b = reconcile_bob_hints(&v);
        assert_eq!(a, b);
        assert_eq!(a.len(), reconcile_total_bytes());
        assert!(reconcile_verify(&v, &v));
    }

    #[test]
    fn test_tri_layer_reconcile_small_noise() {
        let mut rng = rand::thread_rng();
        // Use varied base coefficients spanning the full range to avoid
        // pathological all-zero spectral sums.
        let mut va = [0i16; KYBER_N];
        let mut vb = [0i16; KYBER_N];
        for i in 0..KYBER_N {
            let base = (i * 13) as i16 % KYBER_Q;
            va[i] = base;
            vb[i] = base;
        }
        for i in 0..KYBER_N {
            va[i] += (rng.gen::<u8>() % 12) as i16 - 6;
            vb[i] += (rng.gen::<u8>() % 12) as i16 - 6;
        }
        // For independent noise on varied coefficients, coarse+fine should pass
        // perfectly. Spectral may have rare near-boundary crossings; we count them.
        let mut coarse_mismatches = 0;
        let mut spectral_mismatches = 0;
        for i in 0..KYBER_N {
            let ca = coarse_bucket(va[i]);
            let cb = coarse_bucket(vb[i]);
            if ca.abs_diff(cb) > 1 {
                coarse_mismatches += 1;
            }
        }
        for block_idx in 0..BLOCK_COUNT {
            let mut a = [0i16; BLOCK_SIZE];
            let mut b = [0i16; BLOCK_SIZE];
            for j in 0..BLOCK_SIZE {
                a[j] = va[block_idx * BLOCK_SIZE + j];
                b[j] = vb[block_idx * BLOCK_SIZE + j];
            }
            let s0_a = a[0] as i32 + a[1] as i32 - a[2] as i32 - a[3] as i32;
            let s0_b = b[0] as i32 + b[1] as i32 - b[2] as i32 - b[3] as i32;
            let s1_a = a[0] as i32 + a[2] as i32 - a[1] as i32 - a[3] as i32;
            let s1_b = b[0] as i32 + b[2] as i32 - b[1] as i32 - b[3] as i32;
            let s0_ok = s0_a.signum() == s0_b.signum() || (s0_a.abs() < 64 && s0_b.abs() < 64);
            let s1_ok = s1_a.signum() == s1_b.signum() || (s1_a.abs() < 64 && s1_b.abs() < 64);
            if !s0_ok || !s1_ok {
                spectral_mismatches += 1;
            }
        }
        assert!(
            coarse_mismatches <= 2,
            "Coarse layer had too many mismatches: {} (expected ≤2 with ±6 noise)",
            coarse_mismatches
        );
        assert!(
            spectral_mismatches <= 3,
            "Too many spectral mismatches: {} / {} blocks",
            spectral_mismatches,
            BLOCK_COUNT
        );
    }

    #[test]
    fn test_spectral_catches_correlation() {
        // Create two vectors that differ by a correlated pattern
        // This is the kind of error the spectral layer is designed to catch
        let mut va = [1500i16; KYBER_N];
        let mut vb = [1500i16; KYBER_N];
        for block in 0..BLOCK_COUNT {
            // Add correlated noise: same sign to all 4 coeffs in block
            let noise: i16 = if block % 2 == 0 { 8 } else { -8 };
            for j in 0..BLOCK_SIZE {
                va[block * BLOCK_SIZE + j] += noise;
            }
        }
        // vb gets slightly different correlated noise
        for block in 0..BLOCK_COUNT {
            let noise: i16 = if block % 2 == 0 { 7 } else { -7 };
            for j in 0..BLOCK_SIZE {
                vb[block * BLOCK_SIZE + j] += noise;
            }
        }
        // Even with correlation, coarse+fine should agree, spectral verifies block parity
        let a = reconcile_alice(&va);
        let b = reconcile_bob_hints(&vb);
        let pass = reconcile_verify(&va, &vb);
        assert!(pass, "Spectral verification failed on correlated noise");
        // The packed spectral bytes should match
        let cfb = cofine_bytes();
        assert_eq!(&a[cfb..], &b[cfb..], "Spectral layer mismatch");
    }
}
